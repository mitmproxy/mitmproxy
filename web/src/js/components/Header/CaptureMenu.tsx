import * as React from "react";

CaptureMenu.title = "Capture";

export default function CaptureMenu() {
    //empty component since we don't have an header menu for capture, but we just display the title
    return <></>;
}
