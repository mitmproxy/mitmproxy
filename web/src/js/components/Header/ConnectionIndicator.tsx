import * as React from "react";
import { ConnectionState } from "../../ducks/connection";
import { useAppSelector } from "../../ducks";

export default React.memo(function ConnectionIndicator() {
    const connState = useAppSelector((state) => state.connection.state),
        message = useAppSelector((state) => state.connection.message);

    switch (connState) {
        case ConnectionState.INIT:
            return (
                <span className="connection-indicator init">connecting…</span>
            );
        case ConnectionState.FETCHING:
            return (
                <span className="connection-indicator fetching">
                    fetching data…
                </span>
            );
        case ConnectionState.ESTABLISHED:
            return (
                <span className="connection-indicator established">
                    connected
                </span>
            );
        case ConnectionState.ERROR:
            return (
                <span className="connection-indicator error" title={message}>
                    connection lost
                </span>
            );
        case ConnectionState.OFFLINE:
            return (
                <span className="connection-indicator offline">offline</span>
            );
        default:
            const exhaustiveCheck: never = connState;
            throw "unknown connection state";
    }
});
