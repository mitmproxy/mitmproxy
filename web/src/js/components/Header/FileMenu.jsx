import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import FileChooser from '../common/FileChooser'
import Dropdown, {Divider} from '../common/Dropdown'
import * as flowsActions from '../../ducks/flows'
import * as modalActions from '../../ducks/ui/modal'
import HideInStatic from "../common/HideInStatic";

FileMenu.propTypes = {
    clearFlows: PropTypes.func.isRequired,
    loadFlows: PropTypes.func.isRequired,
    saveFlows: PropTypes.func.isRequired,
}

FileMenu.onNewClick = (e, clearFlows) => {
    e.preventDefault();
    if (confirm('Delete all flows?'))
        clearFlows()
}

export function FileMenu ({clearFlows, loadFlows, saveFlows}) {
     return (
        <Dropdown className="pull-left" btnClass="special" text="mitmproxy">
            <a href="#" onClick={e => FileMenu.onNewClick(e, clearFlows)}>
                <i className="fa fa-fw fa-trash"></i>
                &nbsp;Clear All
            </a>
            <FileChooser
                icon="fa-folder-open"
                text="&nbsp;Open..."
                onOpenFile={file => loadFlows(file)}
            />
            <a href="#" onClick={e =>{ e.preventDefault(); saveFlows();}}>
                <i className="fa fa-fw fa-floppy-o"></i>
                &nbsp;Save...
            </a>

            <HideInStatic>
            <Divider/>
            <a href="http://mitm.it/" target="_blank">
                <i className="fa fa-fw fa-external-link"></i>
                &nbsp;Install Certificates...
            </a>
            </HideInStatic>
        </Dropdown>
    )
}

export default connect(
    null,
    {
        clearFlows: flowsActions.clear,
        loadFlows: flowsActions.upload,
        saveFlows: flowsActions.download,
    }
)(FileMenu)
