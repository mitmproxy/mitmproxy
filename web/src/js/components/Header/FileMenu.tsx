import * as React from "react"
import {useDispatch} from 'react-redux'
import FileChooser from '../common/FileChooser'
import Dropdown, {Divider, MenuItem} from '../common/Dropdown'
import * as flowsActions from '../../ducks/flows'
import HideInStatic from "../common/HideInStatic";
import {useAppSelector} from "../../ducks";


export default React.memo(function FileMenu() {
    const dispatch = useDispatch(), filter = useAppSelector(state => state.flows.filter);
    return (
        <Dropdown className="pull-left special" text="File" options={{"placement": "bottom-start"}}>
            <li>
                <FileChooser
                    icon="fa-folder-open"
                    text="&nbsp;Open..."
                    onClick={
                        // stop event propagation: we must keep the input in DOM for upload to work.
                        e => e.stopPropagation()
                    }
                    onOpenFile={file => {
                        dispatch(flowsActions.upload(file));
                        document.body.click(); // "restart" event propagation
                    }}
                />
            </li>
            <MenuItem onClick={() => location.replace('/flows/dump')}>
                <i className="fa fa-fw fa-floppy-o"/>&nbsp;Save
            </MenuItem>
            <MenuItem onClick={() => location.replace('/flows/dump?filter=' + filter)}>
                <i className="fa fa-fw fa-floppy-o"/>&nbsp;Save filtered
            </MenuItem>
            <MenuItem onClick={() => confirm('Delete all flows?') && dispatch(flowsActions.clear())}>
                <i className="fa fa-fw fa-trash"/>&nbsp;Clear All
            </MenuItem>
            <HideInStatic>
                <Divider/>
                <li>
                    <a href="http://mitm.it/" target="_blank">
                        <i className="fa fa-fw fa-external-link"/>&nbsp;Install Certificates...
                    </a>
                </li>
            </HideInStatic>
        </Dropdown>
    )
});
