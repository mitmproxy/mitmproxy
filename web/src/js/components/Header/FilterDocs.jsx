import React, { Component } from 'react'
import { fetchApi } from "../../utils";


export default class FilterDocs extends Component {

    // @todo move to redux

    static xhr = null
    static doc = null

    constructor(props, context) {
        super(props, context)
        this.state = { doc: FilterDocs.doc }
    }

    componentDidMount() {
        if (!FilterDocs.xhr) {
            FilterDocs.xhr = fetchApi('/filter-help').then(response => response.json())
            FilterDocs.xhr.catch(() => {
                FilterDocs.xhr = null
            })
        }
        if (!this.state.doc) {
            FilterDocs.xhr.then(doc => {
                FilterDocs.doc = doc
                this.setState({ doc })
            })
        }
    }

    render() {
        const { doc } = this.state
        return !doc ? (
            <i className="fa fa-spinner fa-spin"></i>
        ) : (
            <table className="table table-condensed">
                <tbody>
                    {doc.commands.map(cmd => (
                        <tr key={cmd[1]} onClick={e => this.props.selectHandler(cmd[0].split(" ")[0] + " ")}>
                            <td>{cmd[0].replace(' ', '\u00a0')}</td>
                            <td>{cmd[1]}</td>
                        </tr>
                    ))}
                    <tr key="docs-link">
                        <td colSpan="2">
                            <a href="https://mitmproxy.org/docs/latest/concepts-filters/"
                                target="_blank">
                                <i className="fa fa-external-link"></i>
                            &nbsp; mitmproxy docs</a>
                        </td>
                    </tr>
                </tbody>
            </table>
        )
    }
}
