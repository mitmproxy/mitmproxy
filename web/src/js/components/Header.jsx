import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import classnames from 'classnames'
import MainMenu from './Header/MainMenu'
import OptionMenu from './Header/OptionMenu'
import FileMenu from './Header/FileMenu'
import FlowMenu from './Header/FlowMenu'
import {setActiveMenu} from '../ducks/ui/header'
import ConnectionIndicator from "./Header/ConnectionIndicator"
import HideInStatic from './common/HideInStatic'

class Header extends Component {
    static entries = [MainMenu, OptionMenu]

    handleClick(active, e) {
        e.preventDefault()
        this.props.setActiveMenu(active.title)
    }

    render() {
        const { selectedFlowId, activeMenu} = this.props

        let entries = [...Header.entries]
        if(selectedFlowId)
            entries.push(FlowMenu)

        // Make sure to have a fallback in case FlowMenu is selected but we don't have any flows
        // (e.g. because they are all deleted or not yet received)
        const Active = _.find(entries, (e) => e.title == activeMenu) || MainMenu

        return (
            <header>
                <nav className="nav-tabs nav-tabs-lg">
                    <FileMenu/>
                    {entries.map(Entry => (
                        <a key={Entry.title}
                           href="#"
                           className={classnames({ active: Entry === Active})}
                           onClick={e => this.handleClick(Entry, e)}>
                            {Entry.title}
                        </a>
                    ))}
                    <HideInStatic>
                        <ConnectionIndicator/>
                    </HideInStatic>
                </nav>
                <div>
                    <Active/>
                </div>
            </header>
        )
    }
}

export default connect(
    state => ({
        selectedFlowId: state.flows.selected[0],
        activeMenu: state.ui.header.activeMenu,
    }),
    {
        setActiveMenu,
    }
)(Header)
