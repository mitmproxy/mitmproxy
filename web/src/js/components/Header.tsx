import React, { useState } from "react";
import classnames from "classnames";
import StartMenu from "./Header/StartMenu";
import OptionMenu from "./Header/OptionMenu";
import FileMenu from "./Header/FileMenu";
import FlowMenu from "./Header/FlowMenu";
import ConnectionIndicator from "./Header/ConnectionIndicator";
import HideInStatic from "./common/HideInStatic";
import { useAppSelector } from "../ducks";

interface Menu {
    (): JSX.Element;

    title: string;
}

export default function Header() {
    const selectedFlows = useAppSelector((state) =>
            state.flows.selected.filter((id) => id in state.flows.byId)
        ),
        [ActiveMenu, setActiveMenu] = useState<Menu>(() => StartMenu),
        [wasFlowSelected, setWasFlowSelected] = useState(false);

    let entries: Menu[] = [StartMenu, OptionMenu];
    if (selectedFlows.length > 0) {
        if (!wasFlowSelected) {
            setActiveMenu(() => FlowMenu);
            setWasFlowSelected(true);
        }
        entries.push(FlowMenu);
    } else {
        if (wasFlowSelected) {
            setWasFlowSelected(false);
        }
        if (ActiveMenu === FlowMenu) {
            setActiveMenu(() => StartMenu);
        }
    }

    function handleClick(active: Menu, e) {
        e.preventDefault();
        setActiveMenu(() => active);
    }

    return (
        <header>
            <nav className="nav-tabs nav-tabs-lg">
                <FileMenu />
                {entries.map((Entry) => (
                    <a
                        key={Entry.title}
                        href="#"
                        className={classnames({ active: Entry === ActiveMenu })}
                        onClick={(e) => handleClick(Entry, e)}
                    >
                        {Entry.title}
                    </a>
                ))}
                <HideInStatic>
                    <ConnectionIndicator />
                </HideInStatic>
            </nav>
            <div>
                <ActiveMenu />
            </div>
        </header>
    );
}
