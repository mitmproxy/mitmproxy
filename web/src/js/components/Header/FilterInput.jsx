import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import classnames from 'classnames'
import { Key } from '../../utils.js'
import Filt from '../../filt/filt'
import FilterDocs from './FilterDocs'

export default class FilterInput extends Component {

    constructor(props, context) {
        super(props, context)

        // Consider both focus and mouseover for showing/hiding the tooltip,
        // because onBlur of the input is triggered before the click on the tooltip
        // finalized, hiding the tooltip just as the user clicks on it.
        this.state = { value: this.props.value, focus: false, mousefocus: false }

        this.onChange = this.onChange.bind(this)
        this.onFocus = this.onFocus.bind(this)
        this.onBlur = this.onBlur.bind(this)
        this.onKeyDown = this.onKeyDown.bind(this)
        this.onMouseEnter = this.onMouseEnter.bind(this)
        this.onMouseLeave = this.onMouseLeave.bind(this)
        this.selectFilter = this.selectFilter.bind(this)
    }

    componentWillReceiveProps(nextProps) {
        this.setState({ value: nextProps.value })
    }

    isValid(filt) {
        try {
            const str = filt == null ? this.state.value : filt
            if (str) {
                Filt.parse(str)
            }
            return true
        } catch (e) {
            return false
        }
    }

    getDesc() {
        if (!this.state.value) {
            return <FilterDocs selectHandler={this.selectFilter}/>
        }
        try {
            return Filt.parse(this.state.value).desc
        } catch (e) {
            return '' + e
        }
    }

    onChange(e) {
        const value = e.target.value
        this.setState({ value })

        // Only propagate valid filters upwards.
        if (this.isValid(value)) {
            this.props.onChange(value)
        }
    }

    onFocus() {
        this.setState({ focus: true })
    }

    onBlur() {
        this.setState({ focus: false })
    }

    onMouseEnter() {
        this.setState({ mousefocus: true })
    }

    onMouseLeave() {
        this.setState({ mousefocus: false })
    }

    onKeyDown(e) {
        if (e.keyCode === Key.ESC || e.keyCode === Key.ENTER) {
            this.blur()
            // If closed using ESC/ENTER, hide the tooltip.
            this.setState({mousefocus: false})
        }
        e.stopPropagation()
    }

    selectFilter(cmd) {
        this.setState({value: cmd})
        ReactDOM.findDOMNode(this.refs.input).focus()
    }

    blur() {
        ReactDOM.findDOMNode(this.refs.input).blur()
    }

    select() {
        ReactDOM.findDOMNode(this.refs.input).select()
    }

    render() {
        const { type, color, placeholder } = this.props
        const { value, focus, mousefocus } = this.state
        return (
            <div className={classnames('filter-input input-group', { 'has-error': !this.isValid() })}>
                <span className="input-group-addon">
                    <i className={'fa fa-fw fa-' + type} style={{ color }}/>
                </span>
                <input
                    type="text"
                    ref="input"
                    placeholder={placeholder}
                    className="form-control"
                    value={value}
                    onChange={this.onChange}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    onKeyDown={this.onKeyDown}
                />
                {(focus || mousefocus) && (
                    <div className="popover bottom"
                         onMouseEnter={this.onMouseEnter}
                         onMouseLeave={this.onMouseLeave}>
                        <div className="arrow"/>
                        <div className="popover-content">
                        {this.getDesc()}
                        </div>
                    </div>
                )}
            </div>
        )
    }
}
