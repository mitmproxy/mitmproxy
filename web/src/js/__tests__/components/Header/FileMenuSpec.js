import React from 'react'
import renderer from 'react-test-renderer'
import { FileMenu } from '../../../components/Header/FileMenu'

global.confirm = jest.fn( s => true )

describe('FileMenu Component', () => {
    let clearFn = jest.fn(),
        loadFn = jest.fn(),
        saveFn = jest.fn(),
        openModalFn = jest.fn(),
        mockEvent = {
            preventDefault: jest.fn(),
            target: { files: ["foo", "bar "] }
        },
        createNodeMock = () => { return { click: jest.fn() }},
        fileMenu = renderer.create(
            <FileMenu
                clearFlows={clearFn}
                loadFlows={loadFn}
                saveFlows={saveFn}
                openModal={openModalFn}
            />,
            { createNodeMock }),
        tree = fileMenu.toJSON()

    it('should render correctly', () => {
        expect(tree).toMatchSnapshot()
    })

    let ul = tree.children[1]

    it('should clear flows', () => {
        let a = ul.children[0].children[1]
        a.props.onClick(mockEvent)
        expect(mockEvent.preventDefault).toBeCalled()
        expect(clearFn).toBeCalled()
    })

    it('should load flows', () => {
        let fileChooser = ul.children[1].children[1],
            input = fileChooser.children[2]
        input.props.onChange(mockEvent)
        expect(loadFn).toBeCalledWith("foo")
    })

    it('should save flows', () => {
        let a = ul.children[2].children[1]
        a.props.onClick(mockEvent)
        expect(saveFn).toBeCalled()
    })
})
