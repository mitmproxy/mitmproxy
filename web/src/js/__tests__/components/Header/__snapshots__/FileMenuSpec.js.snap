// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileMenu Component should render correctly 1`] = `
<div
  className="dropdown pull-left"
>
  <a
    className="special"
    href="#"
    onClick={[Function]}
  >
    mitmproxy
  </a>
  <ul
    className="dropdown-menu"
    role="menu"
  >
    <li>
       
      <a
        href="#"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-trash"
        />
         Clear All
      </a>
       
    </li>
    <li>
       
      <a
        href="#"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-folder-open"
        />
         Open...
        <input
          className="hidden"
          onChange={[Function]}
          type="file"
        />
      </a>
       
    </li>
    <li>
       
      <a
        href="#"
        onClick={[Function]}
      >
        <i
          className="fa fa-fw fa-floppy-o"
        />
         Save...
      </a>
       
    </li>
    <li>
       
      <hr
        className="divider"
      />
      <a
        href="http://mitm.it/"
        target="_blank"
      >
        <i
          className="fa fa-fw fa-external-link"
        />
         Install Certificates...
      </a>
       
    </li>
  </ul>
</div>
`;
